src/materials/DamagePlasticityStressUpdate.C

```diff
@@ -0,0 +1,680 @@
+                                                   const std::vector<Real> & current_stress_params,
+                                                   const std::vector<Real> & intnl,
+                                                   std::vector<std::vector<Real>> & dintnl) const
+{
+  Real I1_trial = trial_stress_params[0] + trial_stress_params[1] + trial_stress_params[2];
+  Real J2_trial =
+      RankTwoTensor(trial_stress_params[0], trial_stress_params[1], trial_stress_params[2], 0, 0, 0)
+          .secondInvariant();
+  Real invsqrt2J2_trial = 1. / std::sqrt(2. * J2_trial);
+  Real G = 0.5 * (_Eij[0][0] - _Eij[0][1]); // Lame's mu
+  Real K = _Eij[0][1] + 2. * G / 3.;        // Bulk modulus
+  Real C1 = (2. * G * invsqrt2J2_trial);
+  Real C2 = -(I1_trial / 3. * G * invsqrt2J2_trial - 3. * K * _alfa_p);
+  Real C3 = 3. * K * _alfa_p;
+
+  RankTwoTensor dsig = RankTwoTensor(trial_stress_params[0] - current_stress_params[0],
+                                     trial_stress_params[1] - current_stress_params[1],
+                                     trial_stress_params[2] - current_stress_params[2],
+                                     0.,
+                                     0.,
+                                     0.);
+  RankTwoTensor fac = J2_trial < _f_tol ? C3 * RankTwoTensor(1., 1., 1., 0., 0., 0.)
+                                        : RankTwoTensor(C1 * trial_stress_params[0] - C2,
+                                                        C1 * trial_stress_params[1] - C2,
+                                                        C1 * trial_stress_params[2] - C2,
+                                                        0.,
+                                                        0.,
+                                                        0.);
+
+  Real lam = dsig.L2norm() / fac.L2norm();
+
+  std::vector<Real> dlam_dsig(3);
+  for (unsigned i = 0; i < _num_sp; ++i)
+    dlam_dsig[i] = dsig.L2norm() == 0. ? 0.
+                                       : -(trial_stress_params[i] - current_stress_params[i]) /
+                                             (dsig.L2norm() * fac.L2norm());
+
+  std::vector<Real> h(2);
+  hardPotential(current_stress_params, intnl, h);
+  std::vector<std::vector<Real>> dh_dsig(2, std::vector<Real>(3));
+  dhardPotential_dstress(current_stress_params, intnl, dh_dsig);
+  std::vector<std::vector<Real>> dh_dintnl(2, std::vector<Real>(2));
+  dhardPotential_dintnl(current_stress_params, intnl, dh_dintnl);
+
+  for (unsigned i = 0; i < _num_intnl; ++i)
+    for (unsigned j = 0; j < _num_sp; ++j)
+      dintnl[i][j] = dlam_dsig[j] * h[i] + lam * dh_dsig[i][j];
+}
+
+Real
+DamagePlasticityStressUpdate::ft(const std::vector<Real> & intnl) const
+{
+  Real sqrtPhi_t = std::sqrt(1. + _at * (2. + _at) * intnl[0]);
+  if (_zt > sqrtPhi_t / _at)
+    return _ft0 * std::pow(_zt - sqrtPhi_t / _at, (1. - _dt_bt)) * sqrtPhi_t;
+  else
+    return _ft0 * 1.E-6;
```

Comment:
Look into this -- this doesn't seem quite right.

Reply:
If it's a useful constant, do something like:
```
const tol = 1e-6;
return _ft0 * tol;

Reply:
This still needs to be addressed.

Reply:
Wrote a comment per discussion